src/joybus.c

```diff
@@ -57,293 +71,224 @@ static volatile struct SI_regs_s * const SI_regs = (struct SI_regs_s *)0xa480000
 static void * const PIF_RAM = (void *)0x1fc007c0;
 
 /**
- * @brief Wait until the SI is finished with a DMA request
+ * @brief A message to be sent to JoyBus, with its completion callback. 
+ */
+typedef struct {
+    uint64_t input[JOYBUS_BLOCK_DWORDS] __attribute__((aligned(16)));  ///< input message
+    void (*callback)(uint64_t* output);                                ///< callback for completion
+} joybus_msg_t;
+
+#define MAX_JOYBUS_MSGS            8    ///< Maximum number of pending joybus messages
+#define JOYBUS_STATE_IDLE          0    ///< Joybus state: idle (no pending messages)
+#define JOYBUS_STATE_SENDING       1    ///< JoyBus state: sending a message to PIF
+#define JOYBUS_STATE_RECEIVING     2    ///< JoyBus state: receiving a reply from PIF
+
+/** @brief Joybus temporary output buffer */
+static uint64_t joybus_outbuf[JOYBUS_BLOCK_DWORDS] __attribute__((aligned(16)));
+/** @brief Joybus current state (either #JOYBUS_STATE_IDLE, #JOYBUS_STATE_SENDING or #JOYBUS_STATE_RECEIVING) */
+static volatile int joybus_state;
+/** @brief Joybus pending messages (ring buffer) */
+static joybus_msg_t joybus_msgs[MAX_JOYBUS_MSGS];
+/** @brief Pending messages write index */
+static volatile int msgs_widx;
+/** @brief Pending messages read index */
+static volatile int msgs_ridx;
+
+static void si_interrupt(void);
+
+/**
+ * @brief Initialize the joybus subsystem
  */
-static void __SI_DMA_wait( void )
+__attribute__((constructor))
+void __joybus_init(void)
 {
-    while (SI_regs->status & (SI_STATUS_DMA_BUSY | SI_STATUS_IO_BUSY)) ;
+    // FIXME: this constructor requires the __init_interrupts constructor to be
+    // already run. Since we are not 100% sure of how GCC handles constructor
+    // ordering, we call it explicitly here (there's no harm in calling it
+    // multiple times anyway). Revisit this after gathering more information
+    // on constructor ordering.
```

Comment:
Just a note: from my experience currently we call constructors in reverse order with respect to the order for which we provide the object files to the linker. This felt ok to me but I'm not sure on the spec. OTOH haven't experimented with per-translation-unit ordering yet.

Reply:
I think C++ mandates that per-translation-unit order is defined to follow the source code order. By executing the constructors in reverse order, I think we're violating that and possibly causing bugs in C++ code. Anyway, I decided to punt for this PR to avoid the rabbit hole effect